### ISSUE TITLE
Pilosa as an embed library

### ISSUE THREAD

**Issue by mcuadros** [pilosa/pilosa#1057]
### Description

In order to speed up and reduce the moving parts of the ingest, we are thinking onimplement the ingest of pilosa as a embed library, with the goal of do in memory operation instead of use the HTTP interface.

### Success criteria (What criteria will consider this ticket closeable?)

Being able to use Polosa as a Go library inside of my own binary distrubition.

**Comment by jaffee** [comment]
Hey @mcuadros - it has always been our intent for Pilosa to be usable in this way, and in theory you can do it now. `github.com/pilosa/pilosa` is a library, so you could import it and do all kinds of fun things. This kind of usage pattern isn't our focus at the moment, but if you want to give it a try and let us know what kind of problems you run into, we'd be happy to help you sort through them.

It would certainly be helpful to know how you plan to use Pilosa - for example, do you need to run a large cluster, or just a single node? Do you just plan to embed Pilosa for ingest and then use it normally afterwards? What kind of data are you trying to index?

**Comment by mcuadros** [comment]
We are evalutating the project as index for one of our core company project. call [engine](https://github.com/src-d/engine), so the size of the claster is pontecionally any, but we are targeting large scale deployment.

For example our goal deployment is 1PB storage with 100 HDFS/Spark nodes, containing all the source code available in any git repository available on internet. The source of the data is git repositories in HDFS, and our plans is index, all the metadata information from git, sush as, references, commit hashes, filenames, etc. https://github.com/src-d/go-git

So the idea is build a git server with [go-git](https://github.com/src-d/go-git) and pilosa embed, where update the index, is as simple, as just push to the server.

**Comment by jaffee** [comment]
wow... this is very, very cool.

I'm trying to get my head around exactly how this could work. I think that, at least initially, you'll want to look at Pilosa's `/import` endpoint. It still uses HTTP, but aggregates lots of data per-request, so the protocol overhead is not as significant. I advise this instead of embedding because the internal node->node communication in a cluster is also over HTTP, so you would effectively need to write new clustering logic if you decided to embed directly. Or did you have some other ideas about how to handle clustering in the embedded scenario?

I noticed you commented on our GRPC ticket which would probably be a good direction for us to go in reducing protocol overhead as well (both client->node and node->node). I think we could prioritize work on this in the near to medium term if it would be helpful to you.

**Comment by mcuadros** [comment]
Thanks for the fast answers, we are experimienting to see the performance and the resources consumed. I will let you know if we decide to use pilosa on our product. 

About the gRPC, is not a hard requiremnt fo us, is just that I expected to have a gRPC enpoint. The performance vs JSON is must better and the space consumed is lower.

BTW why the gRPC ticket is closed?

**Comment by jaffee** [comment]
We'd decided not to move forward with it immediately for the reasons Cody mentioned in his comment (and no one was actively seeking it) - I think we should look into proteus and/or twirp though, and since you're interested in it I'm going to re-open it.

**Comment by travisturner** [comment]
Hey @mcuadros, were you able to get Pilosa embedded the way you were hoping to? It would be great to understand any problems you may have had, or any feedback you have from the process. Also, if there's anything we can help with let me know. Thanks!

**Comment by gedw99** [comment]
I am interested in this too.

I am actually thinking of using it in a browser via gopherjs and on the server.
It's useful to me because I use a json file based document store instead of a db as it were.

So I prefer to embed directly but keep it loosely coupled by using a control plan / pubsub.
For example inside a browser I often run this stuff inside a webworkers as it's on a different thread.

I have only just started looking into this so can't add too much to the conversation yet.

It's great that your looking for feedback to make it more accessible.

Btw I use bleve in the same way currently

**Comment by jaffee** [comment]
@gedw99 I believe the way pilosa loads and stores bitmap data is incompatible with gopherjs (syscalls like mmap or more generally, filesystem access is not supported in the browser).

There is certainly functionality in Pilosa that could be pulled out and made useful in an in-browser scenario, but disentangling it from the file storage will probably not be a small feat.

**Comment by gedw99** [comment]
its pretty easy to abstract the FS calls. Many packages for doing it.
Then FS calls on a server and a browser will then all work.

A browser has a Service Worker and a Web worker too.
Service Worker is just like a web proxy.
Web Worker is just like a Microservice behind a Message queue
I bring this up because its pretty easy to abstract these too and  thats what i am playing with..

**Comment by jaffee** [comment]
Ah yes, I'd caught wind of some of this: https://www.doxsey.net/blog/extending-gopherjs

So you're storing json documents in a fairly simple file-based store and are looking to add some indexing on top with Pilosa @gedw99? And I take it you need to run everything in the browser?

**Comment by gedw99** [comment]
you go it :)

There recently has been added the ability to compile golang packages to JS packages.
github.com/dave/jsgo 
Any golang pkg gets compiles to JS and deplyoed to the CDN with a hash. It makes it convenient and fast. Eventually that will hopeful do WASM too.  The WASM branch is on the golang branch.

Now there is work going on like what "doxsey" is doing to allow linux golang system calls to JUST WORK in the browser.  
https://github.com/badgerodon
This is the interesting bit. I am playing with it but getting errors still.

TO answer your questions...
I dont want to run everything in the browser. Its more about being able to index json files that are stored locally.  SO when a file is saved in the browser via some function path , you want to index it.
Then when your searching in the browser you can search via the index. 
Trying to keep it KISS  for now. This is a fair bit of breaking edge stuff so just trying out baby steps.

I think first step is to get the same use case running server side where pilossa is embedded, since thats what wil have to happen in the browser. If you can point me to super simple example i can work on it ?

**Comment by jaffee** [comment]
I can point you to some tools for indexing arbitrary structured data (i.e. json) into Pilosa via the [PDK](github.com/pilosa/pdk), but I've just been reviewing what it would look like to embed directly, and I made #1151 because I think we'll need to do that before embedding is straightforward. It isn't strictly necessary, but I think it will help a lot from a maintainability perspective. 

You could instantiate a pilosa.Holder directly and use that since you won't need a lot of the clustering, anti-entropy, etc. stuff. I'll try to work up an example of that later today... should be an interesting exercise.

**Comment by jaffee** [comment]
Just to update this, we've got some work going on to remove dependencies from Pilosa core, which is detailed here #1093 - once that's done, I think it will be easier to explore embedding.

**Comment by gedw99** [comment]
Wow you guys are really moving through this well.

I have gotten most of the gopherjs stuff working. Just waiting on you getting your core to be at a point where I can embed it. 
I also am working on using the exact same embedding approach with Flutter.
Flutter runs on all mobiles, and recently also Desktops.

I will try to put some examples up on GitHub but am in deadline hell right now with other projects.

**Comment by jaffee** [comment]
Just to update here, as of 1.0, it should be quite easy to embed Pilosa in other programs. 

github.com/pilosa/pilosa contains the core functionality, things like the HTTP API, cluster membership, etc. are in subpackages and either call into Pilosa or are injected dependencies.

See the `pilosa/server` package to see how pilosa is typically set up for a server/cluster scenario. Using it embedded will probably just require setting up a pilosa.Server and a pilosa.API and making calls via the pilosa.API.

**Comment by gedw99** [comment]
thanks @jaffee 
wow you got it in done in the end. impressed on following through...!!

**Comment by xenister** [comment]
Hi i'm dev student and i want to use pilosa on my project i wanted to know if there is a doc on how to install and use it embendded, i didn't find anything on the official docs and the comment in this thread a a bit old so i wanted to ask before trying anything.

**Comment by jaffee** [comment]
Hi @xenister 

My comment above from Aug, 2018 is amazingly still quite accurate. The `pilosa/server` subpackage is essentially the wrapper which embeds the core Pilosa functionality and makes it into a "command". Depending on what you want to do, you could actually import `pilosa/server` and use its functionality.

The entry point to the Pilosa binary is at `cmd/pilosa/main.go`. That main package imports `cmd` which has a Root command which sets up all the subcommands of the binary. The main subcommand is "server" which is what actually runs Pilosa (the other subcommands are various utilities like import, inspect, etc). The "server" subcommand is set up in `cmd/server.go` which actually imports the `pilosa/server` subpackage I mentioned above.

Hope that helps. Good luck!

**Comment by kuba--** [comment]
Hey @xenister, in 2018 I embedded pilosa as go package into the `github.com/src-d/go-mysql-server` project. It required a little bit hacking but it's doable. The integration used pure bitmap index and stored mapping (rowID <-> value, columnID <-> record ID) independently into BoltDB.
You can take a look into `go-mysql-server/sql/index/pilosa`, but basically the goal was to create mysql indexes based on bitmaps.

Let me know if you have any questions. Here are some helpful hints:
- It starts from creating an index per db, table (one big index)
https://github.com/src-d/go-mysql-server/blob/62780e17d9e5c17442a57843ec0b367e9e5e1a9b/sql/index/pilosa/driver.go#L645

```go
func (d *Driver) newPilosaIndex(db, table string) (*pilosa.Index, error) {
	name := indexName(db, table)
	path := filepath.Join(d.root, "."+DriverID, name)
	idx, err := pilosa.NewIndex(path, name)
	if err != nil {
		return nil, err
	}
	return idx, nil
}
```

When you have an index you can create fields and start saving data. In mysql we saved data in batches with mapping, concurrently per partition:
 ```
...
	for colID = 0; err == nil; colID++ {
		// commit each batch of objects (pilosa and boltdb)
		if colID%sql.IndexBatchSize == 0 && colID != 0 {
			if err = d.saveBatch(ctx, mapping, colID, b); err != nil {
				return 0, err
			}
		}
...
```
Check `savePilosa` function: https://github.com/src-d/go-mysql-server/blob/62780e17d9e5c17442a57843ec0b367e9e5e1a9b/sql/index/pilosa/driver.go#L538

You can also check concurrent index wrapper:
https://github.com/src-d/go-mysql-server/blob/master/sql/index/pilosa/index.go
Where you have different index lookups.
An essence are lookup iterators:
https://github.com/src-d/go-mysql-server/blob/master/sql/index/pilosa/lookup.go
where all functors let us merge bitmaps.

Basically, to make a long story short - you have to remember about manually opening/closing indexes/fields, lookup rows and call all operations on rows:
https://github.com/src-d/go-mysql-server/blob/62780e17d9e5c17442a57843ec0b367e9e5e1a9b/sql/index/pilosa/lookup.go#L26
